presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -455,7 +528,7 @@ public Page getOutput()
             }
 
             // only flush if we are finishing or the aggregation builder is full
-            if (!finishing && (aggregationBuilder == null || !aggregationBuilder.isFull())) {
+            if (!finishing && (aggregationBuilder == null || !aggregationBuilder.isFull()) && pendingInput == null) {
```

Comment:
This will work poorly if there is a lot of small groups (e.g: we will flush small pages and constantly `resetAggregationBuilder`).

Additional condition here should be that `aggregationBuilder` accumulated enough groups.

Reply:
@findepi @sopel39 Karol, I was wondering about this and was thinking of using `MergingPageOutput` to avoid producing tiny pages. Do you think that will help?

Reply:
Probably not as you would allocate aggregation builders per group. It is probably quite expensive as those are complicated objects